templates/apigee-entity--app.html.twig

```diff
@@ -0,0 +1,54 @@
+{#
+/**
+ * @file
+ * Generic theme implementation to display an Apigee entity.
+ *
+ * Drupal core does not yet provide a generic theme for entities therefore you
+ * have to implement your own entity type specific theme. This will change
+ * when http://dgo.to/2808481 gets fixed which will introduce a generic template
+ * for entities with common preprocess functions and theme suggestions.
+ *
+ * Available variables:
+ * - entity: The entity with limited access to object properties and methods.
+ *   Only method names starting with "get", "has", or "is" and a few common
+ *   methods such as "id", "label", and "bundle" are available. For example:
+ *   - entity.getEntityTypeId() will return the entity type ID.
+ *   - entity.hasField('field_example') returns TRUE if the entity includes
+ *     field_example. (This does not indicate the presence of a value in this
+ *     field.)
+ *   Calling other methods, such as entity.delete(), will result in an exception.
+ *   See \Drupal\apigee_edge\Entity\EdgeEntityInterface for a full list of
+ *   methods.
+ * - label: The title of the entity.
+ * - content: All rendered field items. Use {{ content }} to print them all,
+ *   or print a subset such as {{ content.field_example }}. Use
+ *   {{ content|without('field_example') }} to temporarily suppress the printing
+ *   of a given child element.
+ * - url: Direct URL of the current entity.
+ * - attributes: HTML attributes for the containing element.
+ * - title_attributes: Same as attributes, except applied to the main title
+ *   tag that appears in the template.
+ * - content_attributes: Same as attributes, except applied to the main
+ *   content tag that appears in the template.
+ * - title_prefix: Additional output populated by modules, intended to be
+ *   displayed in front of the main title tag that appears in the template.
+ * - title_suffix: Additional output populated by modules, intended to be
+ *   displayed after the main title tag that appears in the template.
+ * - view_mode: View mode; for example, "teaser" or "full".
+ *
+ * @see \Drupal\apigee_edge\Entity\EdgeEntityViewBuilder::getBuildDefaults()
+ * @see \Drupal\Core\Entity\EntityViewBuilder::getBuildDefaults()
+ * @see template_preprocess_apigee_entity()
+ * @see https://www.drupal.org/project/drupal/issues/2808481
+ *
+ * @ingroup themeable
+ */
+#}
+
+{% extends "apigee-entity.html.twig" %}
+{% block apigee_entity %}
+  {% if entity.getEntityTypeId in ['developer_app', 'team_app'] %}
+    {% set attributes = attributes.addClass('apigee-entity--app', 'apigee-entity--app--view-mode-' ~ view_mode|clean_class) %}
+  {% endif %}
+  {{ parent() }}
```

Comment:
I think using Twig template inheritance here is ok, and the general idea of adding the app-specific here is exactly what I was getting at in a previous comment 👏, but this can/should be done differently.  `extends` is the strictest of all methods of doing this, and `parent()` suggests a really tightly coupled relation between the parent and child template, which is not necessarily going to be the case here.  If the structure of the base template is changed, and `{% block apigee_entity %}` is removed, this will throw a fatal error, which the person modifying the base template might not become aware of until after the change is pushed to production and someone happens to look at a page where an app renders and gets a WSOD. 

All we really need here is to use the main template and pass in additional classes, so structuring the base template to allow for that more easily would be ideal.  I can help you with this.

Reply:
RE: 
> I can help you with this.

There are 2 parts to this, as we have to make classes extendable from the base template.  I will post that in context in the apigee-entity.html.twig source, but here is this part:

```twig
{% if entity.getEntityTypeId in ['developer_app', 'team_app'] %}
  {% set classes = [
    'apigee-entity--app',
    'apigee-entity--app--view-mode-' ~ view_mode|clean_class
  ] %}
{% endif %}

{% include 'apigee-entity.html.twig' %}
```

Here we:

- Don't make any assumptions about the base template structure, we just use it directly.
- Add additional classes, which get passed through to the base template, and merged there.  More explicitly, the is equivalent to: `{% include 'apigee-entity.html.twig' with: { classes: classes }  %}`